In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import statsmodels.api as sm
import statistics as stat
from sklearn.linear_model import LinearRegression 
import math

In [2]:
datafile = '../../dataset/ch07/bookstore_sales.txt'
bookstore = pd.read_csv(datafile, sep='\t')

In [3]:
bookstore.info

<bound method DataFrame.info of    Number of Sales People Working  Sales (in $1000)
0                               2                10
1                               3                11
2                               7                13
3                               9                14
4                              10                18
5                              10                20
6                              12                20
7                              15                22
8                              16                22
9                              20                26>

In [4]:
len(bookstore)

10

In [5]:
bookstore.columns

Index(['Number of Sales People Working', 'Sales (in $1000)'], dtype='object')

In [6]:
bookstore.head()

Number of Sales People Working  Sales (in $1000)
0                               2                10
1                               3                11
2                               7                13
3                               9                14
4                              10                18

*Prolem 5*


a) Find the slope estimate, b1.

b) What does it mean, in this context?

c) Find the intercept, b0.

d) What does it mean, in this context? Is it meaningful?

e) Write down the equation that predicts Sales from Number of
Sales People Working.

f) If 18 people are working, what Sales do you predict?

g) If sales are actually $25,000, what is the value of the residual?

h) Have we overestimated or underestimated the sales?

In [30]:
# teacher's code by scikit-learn
linfit = LinearRegression()
x = np.array(bookstore['Number of Sales People Working']).reshape(-1, 1)
y = np.array(bookstore['Sales (in $1000)'])
linfit.fit(x, y)
print(f'b1 = {linfit.coef_[0]:.3f}')
print(f'b1 = {linfit.intercept_:.3f}')
print(f'Predicted sales for 18 salespeople: {linfit.predict(np.array([18]).reshape(1,1))[0]:.3f}')
e = y - linfit.predict(x)
print('residues:')
print((np.max(np.abs(e))), np.min(np.abs(e)))

b1 = 0.913
b1 = 8.101
Predicted sales for 18 salespeople: 24.542
residues:
2.7653631284916216 0.07262569832402299


In [9]:
## 5a) Find the slope estimate, b1.
r = 0.965
sd_y = 5.34
sd_x = 5.64
b_1 = sd_y / sd_x * r
round (b_1, 3)

0.914

**5b) What does it mean, in this context?**


Answer: The slope, 0.914, says that a "sales in 1,000 dollars" with one more salesperson working can be expected, on average, to have 0.914 more of 1000 dollars sales (which is $914).

**Correction:** The slope is a fitted proportionality between "sales in 1,000 dollars" (y) and "number of salespeople" (x).  Roughly speaking, the manager may expect that adding 1 to x (meaning hiring one more salesperson) would lead to an increase of 0.914 in y (meaning that the sales goes up by approximately $914). 


In [16]:
## c) Find the intercept, b0.

y_bar = 17.6
x_bar = 10.4
r = 0.965
sd_y = 5.34
sd_x = 5.64
b_1 = sd_y / sd_x * r
b_0 = y_bar - b_1 * x_bar
round (b_0, 3)

8.098

**5d) What does it mean, in this context? Is it meaningful?**

Answer: The intercept, 8.098, says that the model predicts that even a bookstore with no salespeople would have, on average, about 8.098 of 1,000 dollars of sales, i.e., $8,098.  This is not meaningful here, because we do not expect a no-man bookstore would have any sales.

**5e) Write down the equation that predicts Sales from Number of Sales People Working.**


Answer: $\hat{y} = b_0 + b_1 x$, 
where $\hat{y}$ represents the predicted sales, and $x$ stands for the number of salespeople at the bookstore.  Numerically,

$\hat{y} = 8.098 + 0.914 x$

In [17]:
## 5f) If 18 people are working, what Sales do you predict?

# y_bar = 17.6
# x_bar = 10.4
# r = 0.965
# sd_y = 5.34
# sd_x = 5.64
# b_1 = sd_y / sd_x * r
# b_0 = y_bar - b_1 * x_bar

# r = 0.965
# sd_y = 5.34
# sd_x = 5.64
# b_1 = sd_y / sd_x * r

b0 = 8.098
b1 = 0.914
x = 18

y_hat = b_0 + b_1 * x

print (f"If 18 people are working, I predict {y_hat:.3f} of sales.")

If 18 people are working, I predict 24.544 of sales.


In [24]:
## 5g) If sales are actually $25,000, what is the value of the residual?
residual = 25000 - 24.544 * 1000
residual

456.0

In [27]:
## 5h) not meant to get you compute the R^2

r_2 = 0.93
r_error_term = 1 - (r_2)
print (f"The error term is {r_error_term*100:.2f} precent")

The error term is 7.00 precent


**5h)**


Answer: So 7 precent of the variability in toal sales has been left in the residuals, therefore, there is 93 precent of association between sales and number of sales people working.

**Correction**: the residual of $456 is the difference between actual sales and predicted sales.  Predicated sales is lower than the actual sales.  Therefore, we have underestimated the sales.

## Problem 9

a) What are the units of the residuals?

b) Which residual contributes the most to the sum that was
minimized according to the least squares criterion to find
this regression?

c) Which residual contributes least to that sum?

* a) What are the units of the residuals?


Answer: The units of the residuals are in 1000 dollars.

* b)Which residual contributes the most to the sum that was
minimized according to the least squares criterion to find
this regression?


Answer: $2.77$ residual contributes the most to the sum that was minimized according to the least squares criterion to find this regression.

The next largest residual is $-2.32$.  Although this is negative (meaning overestimation), its absolute value (or just say its magnitude) is the second largest.

* c) Which residual contributes least to that sum?


Answer: 0.07 residual contributes the least to that sum.

**11 For the regression model for the bookstore of Exercise 5, what is the value of R2 and what does it mean?**


Answer: From the regression model for the bookstore of Exercise 5, the R2 value is 0.93 and it means there is 93 precent of association between sales and number of sales people working.

**Correction**: The meaning of $R^2$ is the percentage of the variation (think about variance and standard deviation) in the response variable (y) as accouted for by the variation in the control variable (x).

In this case, 93% of the variation in the sales can be accounted for by the number of salespeople.

In [31]:
## 11 Have we overestimated or underestimated the sales?
r = 0.965
r_2 = 0.965 * 0.965
round (r_2, 2)

0.93

## Some scatterplots

**27. Tell what each of the residual plots below indicates about the appropriateness of the linear model that was fit to the data.**

**Notes**: We need to check:
1. straight enough?
2. outlier(s)?
3. thickening?
In the residual plot, we need to check the data points are scattered randomly, fitting to a horizontal line through 0, and without any outliers or thickening.

Answer: The residual plot (b) shows the appropriateness of the linear model that was fit to the data, because the plot (b) does not have non-linear signal, no obvious outliers , and the plot does not thicken.

(a) This looks like a random plot with no outliers or thickening.  Technically, the adopted linear regression analysis is appropriate.

(b) The plot shows an evident curvature, indicating a nonlinear association between x and y.  Linear regression analysis is NOT appropriate in this case.

(c) We see thickening in this plot (heteroskedasticity), so we need to adopt more advanced (beyond the coverage of this course or book) to analyze the data.

**28 Tell what each of the residual plots below indicates about the appropriateness of the linear model that was fit to the data.**

Answer: The residual plot (a) shows the appropriateness of the linear model that was fit to the data, because the plot (a) does not have obvious outliers, the plot does not thicken and no non-linear signal.

Plot (c) shows the residues after an appropriate linear regression.  Problem with plot (a) is the evident curvature, and the issue with plot (b) is the thickening.